File: python/src/contriwork_market_data/_registry.py
"""AdapterRegistry — maps ``market`` string to ordered adapter chain.

The registry is constructed by the caller (typically the application
wiring code in their DI container or factory function). Default chains for
well-known markets ship in
:func:`contriwork_market_data.default_chains` once the concrete adapters
are imported.
"""

from __future__ import annotations

from collections.abc import Iterable

from ._adapter import Adapter

__all__ = ["AdapterRegistry"]


class AdapterRegistry:
    def __init__(self, chains: dict[str, Iterable[Adapter]] | None = None) -> None:
        self._chains: dict[str, tuple[Adapter, ...]] = {}
        for market, adapters in (chains or {}).items():
            self._chains[market] = tuple(adapters)

    def chain_for(self, market: str) -> tuple[Adapter, ...]:
        """Return the ordered adapter chain for ``market``; empty if unknown."""
        return self._chains.get(market, ())

    def register(self, market: str, adapters: Iterable[Adapter]) -> None:
        self._chains[market] = tuple(adapters)

    def markets(self) -> tuple[str, ...]:
        return tuple(self._chains.keys())

    def adapters(self) -> tuple[Adapter, ...]:
        """Return every unique adapter across all chains, in first-seen order."""
        seen: set[str] = set()
        result: list[Adapter] = []
        for adapters in self._chains.values():
            for adapter in adapters:
                if adapter.adapter_id not in seen:
                    seen.add(adapter.adapter_id)
                    result.append(adapter)
        return tuple(result)

File: typescript/src/registry.ts
/**
 * AdapterRegistry — maps a `market` string to an ordered adapter chain.
 * CONTRACT.md §4.
 */
import type { MarketDataAdapter } from "./adapter.js";

/** Market-to-chain registry. */
export class AdapterRegistry {
  private readonly chains = new Map<string, readonly MarketDataAdapter[]>();

  public constructor(
    chains?: Readonly<Record<string, readonly MarketDataAdapter[]>>,
  ) {
    if (chains !== undefined) {
      for (const [market, adapters] of Object.entries(chains)) {
        this.chains.set(market, [...adapters]);
      }
    }
  }

  /** Register or replace the adapter chain for a market. */
  public register(
    market: string,
    adapters: readonly MarketDataAdapter[],
  ): void {
    this.chains.set(market, [...adapters]);
  }

  /** Get the ordered adapter chain for a market; empty when unknown. */
  public chainFor(market: string): readonly MarketDataAdapter[] {
    return this.chains.get(market) ?? [];
  }

  /** All registered market strings. */
  public markets(): readonly string[] {
    return [...this.chains.keys()];
  }
}
